pkg/receive/handler.go

```diff
@@ -180,8 +191,82 @@ func (h *Handler) receive(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	if err := h.receiver.Receive(&wreq); err != nil {
+	tenant := r.Header.Get(tenantHeader)
+	local, err := h.forward(r.Context(), tenant, &wreq)
+	if err != nil {
 		http.Error(w, err.Error(), http.StatusBadRequest)
 		return
 	}
+	// There may be no WriteRequest destined for the local node.
+	if local != nil {
+		if err := h.receiver.Receive(local); err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+	}
+}
+
+// forward accepts a write request, batches its time series by
+// corresponding endpoint, and forwards them in parallel. It returns a write
+// request containing only the time series that correspond to
+// local handler.
+func (h *Handler) forward(ctx context.Context, tenant string, wreq *prompb.WriteRequest) (*prompb.WriteRequest, error) {
+	wreqs := make(map[string]*prompb.WriteRequest)
+	for i := range wreq.Timeseries {
+		endpoint, err := h.hashring.Get(tenant, &wreq.Timeseries[i])
+		if err != nil {
+			return nil, err
+		}
+		if _, ok := wreqs[endpoint]; !ok {
+			wreqs[endpoint] = &prompb.WriteRequest{}
+		}
+		wr := wreqs[endpoint]
+		wr.Timeseries = append(wr.Timeseries, wreq.Timeseries[i])
+	}
+
+	ec := make(chan error)
+	defer close(ec)
+	var n int
+	var local *prompb.WriteRequest
+	for endpoint := range wreqs {
+		if endpoint == h.options.Endpoint {
+			local = wreqs[endpoint]
+			continue
+		}
+		n++
+		go func(endpoint string) {
+			buf, err := proto.Marshal(wreqs[endpoint])
+			if err != nil {
+				level.Error(h.logger).Log("msg", "proto marshal error", "err", err, "endpoint", endpoint)
+				ec <- err
+				return
+			}
+			req, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(snappy.Encode(nil, buf)))
+			if err != nil {
+				level.Error(h.logger).Log("msg", "create request error", "err", err, "endpoint", endpoint)
+				ec <- err
+				return
+			}
+			req.Header.Add(tenantHeader, tenant)
+			if _, err := http.DefaultClient.Do(req.WithContext(ctx)); err != nil {
+				level.Error(h.logger).Log("msg", "forward request error", "err", err, "endpoint", endpoint)
+				ec <- err
+				return
+			}
+			ec <- nil
+		}(endpoint)
+	}
+
+	var errs error
+	for ; n > 0; n-- {
```

Comment:
Couldn't we utilize a sync.WaitGroup here? I might be missing something more obvious though.
https://gobyexample.com/waitgroups

Reply:
We could and it originally did but it makes it more complex in reality. Plus it introduces an unnecessary second synchronization mechanism (the first being the error chan).

Reply:
It complicated things because in order to do `wg.Wait()` we need to put the error channel receiver in a new go routine (otherwise the waitgroups wouldn’t finish (unless the chan was buffered by n)) and then we need to close the chan in a different go routine in a way that doesn’t race with the data that is shared, i.e. the collected errs. Keeping the collection in the main goroutine simplifies the code IMO and spawns fewer gorouines.